### PR TITLE
Scalar Quantization

### PR DESCRIPTION
**Pull request by abdelr**
## What's being changed:

- This PR adds a new scalar quantizer to be used together with HNSW
- Provides 1/4 memory usage, faster import times, and often a better recall/qps curves (dependent on vectors)
- It is more robust that binary quantization in the case where the model was not training to be away of binary quantized loss
- Uses SIMD implemented scalar distance functions for amd64 and arm64 architectures (supporting dot product, cosine, and l2 distance metrics) https://github.com/weaviate/weaviate/pull/5147 https://github.com/weaviate/weaviate/pull/4756 https://github.com/weaviate/weaviate/pull/5095 https://github.com/weaviate/weaviate/pull/5077
- The bounds of scalar quantization depend on the vectors, because of this scalar quantization must be enabled after loading data, or async indexing must be enabled to automatically enable scalar quantization (in a similar way to product quantization).
- Currently not supported with the flat index

## Configuration options
```json
{
  "vectorIndexConfig": {
    "sq": {
      "enabled": true,
      "trainingLimit": 10000,
      "rescoreLimit": 20 
  }
}
```

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/repos/db/vector/hnsw/compress.go [hunk 1]
```diff
@@ -86,7 +86,45 @@ func (h *hnsw) compress(cfg ent.UserConfig) error {
...
 		}
-		h.commitLog.AddPQ(h.compressor.ExposeFields())
+		h.commitLog.AddCompression(h.compressor.ExposeFields())
+	} else if cfg.SQ.Enabled {
+		if h.isEmpty() {
+			return errors.New("Compress command cannot be executed before inserting some data. Please, insert your data first.")
+		}
+
+		cleanData := make([][]float32, 0, len(data))
+		for i := range data {
```
Comment (trengrj) [inline adapters/repos/db/vector/hnsw/compress.go:96]: Are we going to have a `trainingLimit` for SQ as well?

File: adapters/repos/db/vector/hnsw/startup.go [hunk 2]
```diff
@@ -132,28 +132,47 @@ func (h *hnsw) restoreFromDisk() error {
...
+				data.B,
+				data.Dimensions,
 				h.store,
 				h.allocChecker,
 			)
 			if err != nil {
 				return errors.Wrap(err, "Restoring compressed data.")
 			}
+		} else {
+			return errors.New("unsopported type while loading compression data")
```
Comment (trengrj) [inline adapters/repos/db/vector/hnsw/startup.go:175]:
```suggestion
			return errors.New("unsupported type while loading compression data")
```

File: adapters/repos/db/vector/hnsw/compress.go (hunk 1 above)
Comment (abdelr) [reply adapters/repos/db/vector/hnsw/compress.go]: Yes, already add the check for it. Thanks for pointing this out.

File: adapters/repos/db/vector/hnsw/search.go [hunk 3]
```diff
@@ -563,7 +549,12 @@ func (h *hnsw) knnSearchByVector(searchVec []float32, k int,
 		return nil, nil, errors.Wrapf(err, "knn search: search layer at level %d", 0)
 	}
 
-	if h.shouldRescore() {
+	if h.shouldRescore() && h.sqConfig.RescoreLimit >= k {
```
Comment (trengrj) [inline adapters/repos/db/vector/hnsw/search.go:552]: @abdelr will this still work with PQ?

**Review by trengrj — DISMISSED** [review]

File: adapters/repos/db/vector/hnsw/deserializer.go [hunk 4]
```diff
@@ -143,6 +144,9 @@ func (d *Deserializer) Do(fd *bufio.Reader, initialState *DeserializationResult,
 		case AddPQ:
 			err = d.ReadPQ(fd, out)
 			readThisRound = 9
+		case AddSQ:
+			err = d.ReadSQ(fd, out)
+			readThisRound = 9
```
Comment (trengrj) [inline adapters/repos/db/vector/hnsw/deserializer.go:149]: @abdelr shouldn't this be 10 as `toWrite` is 11 for SQ?

Comment (abdelr) [reply adapters/repos/db/vector/hnsw/deserializer.go]: Fixed

**Review by trengrj — DISMISSED** [review]

File: adapters/repos/db/vector/compressionhelpers/compression.go [hunk 5]
```diff
@@ -134,15 +134,15 @@ func (compressor *quantizedVectorsCompressor[T]) DistanceBetweenCompressedVector
 	return dist, err
 }
 
-func (compressor *quantizedVectorsCompressor[T]) DistanceBetweenCompressedAndUncompressedVectorsFromID(ctx context.Context, id uint64, vector []float32) (float32, error) {
+/*func (compressor *quantizedVectorsCompressor[T]) DistanceBetweenCompressedAndUncompressedVectorsFromID(ctx context.Context, id uint64, vector []float32) (float32, error) {
```
Comment (antas-marcin) [inline adapters/repos/db/vector/compressionhelpers/compression.go:137]: shouldn't this commented code be removed?

Comment (abdelr) [reply adapters/repos/db/vector/compressionhelpers/compression.go]: Sure, done already, thanks @antas-marcin

**Review by antas-marcin — APPROVED** [review]

**Pull request merged**